process/peer/process.go

```diff
@@ -0,0 +1,288 @@
+	if err != nil {
+		return err
+	}
+
+	err  = p.updateValidatorInfo(consensusGroup, previousHeader.GetPubKeysBitmap(), previousHeader.GetShardID())
+	if err != nil {
+		return err
+	}
+
+	if header.GetShardID() == sharding.MetachainShardId {
+		err = p.updateShardDataPeerState(header)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *peerProcess) updateShardDataPeerState(header data.HeaderHandler) error {
+	metaHeader, ok := header.(*block.MetaBlock)
+	if !ok {
+		return process.ErrInvalidMetaHeader
+	}
+
+	for _, h := range metaHeader.ShardInfo {
+		shardHeaderBytes, err := p.shardHeaderStorage.Get(h.HeaderHash)
+		if err != nil {
+			return err
+		}
+
+		shardHeader := &block.Header{}
+		err = p.marshalizer.Unmarshal(shardHeader, shardHeaderBytes)
+		if err != nil {
+			return err
+		}
+
+		shardConsensus, err := p.nodesCoordinator.ComputeValidatorsGroup(shardHeader.GetRandSeed(), shardHeader.GetRound(), shardHeader.GetShardID())
+		if err != nil {
+			return err
+		}
+
+		err = p.updateValidatorInfo(shardConsensus, shardHeader.GetPubKeysBitmap(), shardHeader.GetShardID())
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *peerProcess) processNode(node *sharding.InitialNode) error {
+	if !p.IsNodeValid(node) {
+		return process.ErrInvalidInitialNodesState
+	}
+
+	peerAccount, err := p.generatePeerAccount(node)
+	if err != nil {
+		return err
+	}
+
+	err = p.savePeerAccountData(peerAccount, node)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *peerProcess) generatePeerAccount(node *sharding.InitialNode) (*state.PeerAccount, error) {
+	address, err := p.adrConv.CreateAddressFromHex(node.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	peerAdapter, err := p.getPeerAdapter(node.AssignedShard())
+	if err != nil {
+		return nil, err
+	}
+
+	acc, err := peerAdapter.GetAccountWithJournal(address)
+	if err != nil {
+		return nil, err
+	}
+
+	peerAccount, ok := acc.(*state.PeerAccount)
+	if !ok {
+		return nil, process.ErrInvalidPeerAccount
+	}
+
+	return peerAccount, nil
+}
+
+func (p *peerProcess) savePeerAccountData(peerAccount *state.PeerAccount, data *sharding.InitialNode) error {
+	err := peerAccount.SetAddressWithJournal([]byte(data.Address))
+	if err != nil {
+		return err
+	}
+
+	err = peerAccount.SetSchnorrPublicKeyWithJournal([]byte(data.Address))
+	if err != nil {
+		return err
+	}
+
+	err = peerAccount.SetBLSPublicKeyWithJournal([]byte(data.PubKey))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *peerProcess) updateValidatorInfo(validatorList []sharding.Validator, signingBitmap []byte, shardId uint32) error {
+	lenValidators := len(validatorList)
+	for i := 0; i < lenValidators; i++ {
+		peerAcc, err := p.getPeerAccount(validatorList[i].Address(), shardId)
+		if err != nil {
+			return err
+		}
+
+		isLeader := i == 0
+		validatorSigned := (signingBitmap[i/8] & (1 << (uint16(i) % 8))) != 0
+		if isLeader && validatorSigned {
```

Comment:
should be done with a switch

Reply:
@AdoAdoAdo I added a function and a switch here. But we talked about using a mask, instead I left the logic with if's since it seems more easy to read. But if you don't agree we can change this